File: Algorithms/C++/Medium/369_Plus_One_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* plusOne(ListNode* head) {
        if (head == NULL) {
            return NULL;
        }

        ListNode* reversedHead = reverse(head);
        addOne(reversedHead);

        return reverse(reversedHead);
    }
private:
    ListNode* reverse(ListNode* head) {
        if (head == NULL) {
            return NULL;
        }

        ListNode* prev = NULL;
        while (head != NULL) {
            ListNode* tmp = head->next;
            head->next = prev;
            prev = head;
            head = tmp;
        }

        return prev;
    }

    void addOne(ListNode* head) {
        if (head == NULL) {
            return;
        }

        int carry = 1;
        ListNode* prev = NULL;
        while (head != NULL) {
            head->val += carry;

            carry = head->val / 10;
            head->val %= 10;

            prev = head;
            head = head->next;
        }

        if (carry != 0) {
            ListNode* newNode = new ListNode(carry);
            prev->next = newNode;
        }
    }
};

File: Algorithms/Swift/Medium/369_Plus_One_Linked_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func plusOne(head: ListNode?) -> ListNode? {
        if head == nil {
            return nil
        }
        
        let reversedHead = reverse(head)
        addOne(reversedHead)
        
        return reverse(reversedHead)
    }
    
    private func reverse(head: ListNode?) -> ListNode? {
        if head == nil {
            return nil
        }
        
        var previousNode: ListNode!
        var node = head
        while node != nil {
            let tmp = node?.next
            node?.next = previousNode
            previousNode = node
            node = tmp
        }
        
        return previousNode
    }
    
    private func addOne(head: ListNode?) {
        if head == nil {
            return
        }
        
        var carry = 1
        var node = head
        var previousNode: ListNode?
        while node != nil {
            node!.val += carry
            
            carry = node!.val / 10
            node!.val %= 10
            
            previousNode = node
            node = node?.next
        }
        
        if carry != 0 {
            previousNode?.next = ListNode(carry)
        }
    }
}
